api/src/services/user-service.ts

```diff
@@ -227,4 +234,60 @@ export class UserService extends DBService {
       throw new ApiExecuteSQLError('Failed to insert user system roles');
     }
   }
+
+  /**
+   * Update a system user's record with the latest information from a verified Keycloak token.
+   *
+   * Note: Does nothing if the user is an internal database user.
+   *
+   * @param {KeycloakUserInformation} keycloakUserInformation
+   * @return {*}  {(Promise<number | null>)} the system user id, if a matching system user record was found and updated,
+   * otherwise null.
+   * @memberof UserService
+   */
+  async updateSystemUserInformation(keycloakUserInformation: KeycloakUserInformation): Promise<number | null> {
+    let values;
+
+    if (isDatabaseUserInformation(keycloakUserInformation)) {
+      // Don't patch internal database user records
+      return null;
+    }
+
+    // We don't yet know at this point what kind of token was used (idir vs bceid basic, etc).
+    // Determine which type it is, and parse the information into a generic structure
+    if (isIdirUserInformation(keycloakUserInformation)) {
+      values = {
+        user_guid: keycloakUserInformation.idir_user_guid,
+        user_identifier: keycloakUserInformation.idir_username,
+        user_identity_source: SYSTEM_IDENTITY_SOURCE.IDIR,
+        display_name: keycloakUserInformation.display_name,
+        email: keycloakUserInformation.email,
+        given_name: keycloakUserInformation.given_name,
+        family_name: keycloakUserInformation.family_name
+      };
+    } else if (isBceidBusinessUserInformation(keycloakUserInformation)) {
+      values = {
+        user_guid: keycloakUserInformation.bceid_user_guid,
+        user_identifier: keycloakUserInformation.bceid_username,
+        user_identity_source: SYSTEM_IDENTITY_SOURCE.BCEID_BUSINESS,
+        display_name: keycloakUserInformation.display_name,
+        email: keycloakUserInformation.email,
+        given_name: keycloakUserInformation.given_name,
+        family_name: keycloakUserInformation.family_name,
+        agency: keycloakUserInformation.bceid_business_name
+      };
+    } else {
```

Comment:
is it safe to assume we'll hit one of these types?

Reply:
It should be, if the upstream calls are being typesafe.

If you've generated the `keycloakUserInformation` using the function (which just calls the zod schema on the raw token), it should be safe to assume it will match one of these here.

Reply:
ok great